src/client/dfs/dfs.c

```diff
@@ -347,15 +348,20 @@ fetch_entry(daos_handle_t oh, daos_handle_t th, const char *name,
 		if (rc) {
 			D_ERROR("Failed to fetch entry %s " DF_RC "\n", name,
 				DP_RC(rc));
+			D_FREE(value);
 			D_GOTO(out, rc = daos_der2errno(rc));
 		}
 
-		sym_len = sg_iovs[0].iov_len;
-		if (sym_len != 0) {
-			D_ASSERT(value);
-			D_STRNDUP(entry->value, value, PATH_MAX - 1);
-			if (entry->value == NULL)
-				D_GOTO(out, rc = ENOMEM);
+		entry->value_len = sg_iovs[0].iov_len;
+
+		if (entry->value_len != 0) {
+			/* Return value here, and allow the caller to truncate
+			 * the buffer if they want to
+			 */
+			entry->value = value;
+		} else {
+			D_ERROR("Failed to load value for symlink\n");
```

Comment:
I wasn't sure about this, the code before would silently allow it, but then the caller would check and return EIO rather than assert, so I think it was probably to protect against incorrect object format.  I agree this could set rc, and if it did that then a couple of other checks could be changed to D_ASSERTS()